File: backend/app/models/task.py
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, Boolean, Date
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.database import Base


class Task(Base):
    __tablename__ = "tasks"
    
    id = Column(Integer, primary_key=True, index=True)
    company_id = Column(Integer, ForeignKey("companies.id"), nullable=False, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    
    status = Column(String(50), default="Pending")  # Pending, In Progress, Completed
    priority = Column(String(20), default="medium")  # low, medium, high
    
    # Due date
    due_date = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)
    
    # Related entity (Lead or Client)
    lead_id = Column(Integer, ForeignKey("leads.id"), nullable=True)
    client_id = Column(Integer, ForeignKey("clients.id"), nullable=True)
    
    # Assignment
    assigned_to_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    assigned_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    
    # Manager-assigned flag
    is_manager_assigned = Column(Boolean, default=False)
    
    # Timestamps
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())
    
    # Relationships
    company = relationship("Company", backref="tasks")
    lead = relationship("Lead", back_populates="tasks")
    client = relationship("Client", back_populates="tasks")
    assigned_to = relationship("User", foreign_keys=[assigned_to_id], back_populates="tasks_assigned")
    assigned_by = relationship("User", foreign_keys=[assigned_by_id], back_populates="tasks_created")

File: flutter_app/lib/data/models/task.dart
class Task {
  final int id;
  final String title;
  final String? description;
  final String status;
  final String priority;
  final DateTime? dueDate;
  final DateTime? completedAt;
  final int? assignedToId;
  final String? assignedToName;
  final int? assignedById;
  final bool isManagerAssigned;
  final int? leadId;
  final int? clientId;
  final DateTime? createdAt;

  const Task({
    required this.id,
    required this.title,
    this.description,
    required this.status,
    required this.priority,
    this.dueDate,
    this.completedAt,
    this.assignedToId,
    this.assignedToName,
    this.assignedById,
    this.isManagerAssigned = false,
    this.leadId,
    this.clientId,
    this.createdAt,
  });

  bool get isCompleted => status == 'Completed';
  bool get isOverdue =>
      dueDate != null && !isCompleted && dueDate!.isBefore(DateTime.now());

  factory Task.fromJson(Map<String, dynamic> json) => Task(
        id: json['id'] as int,
        title: json['title'] as String,
        description: json['description'] as String?,
        status: json['status']?.toString() ?? 'Pending',
        priority: json['priority']?.toString() ?? 'Medium',
        dueDate: json['due_date'] != null
            ? DateTime.tryParse(json['due_date'])
            : null,
        completedAt: json['completed_at'] != null
            ? DateTime.tryParse(json['completed_at'])
            : null,
        assignedToId: json['assigned_to_id'] as int?,
        assignedToName: json['assigned_to_name'] as String?,
        assignedById: json['assigned_by_id'] as int?,
        isManagerAssigned: json['is_manager_assigned'] == true,
        leadId: json['lead_id'] as int?,
        clientId: json['client_id'] as int?,
        createdAt: json['created_at'] != null
            ? DateTime.tryParse(json['created_at'])
            : null,
      );
}
